src/Models/Refund.php

```diff
@@ -6,191 +6,116 @@
+use Tipoff\Authorization\Models\User;
+use Tipoff\Payments\Models\Payment;
+use Tipoff\Refunds\Enums\RefundMethod;
 use Tipoff\Refunds\Notifications\RefundConfirmation;
+use Tipoff\Refunds\Services\IssueRefund;
+use Tipoff\Support\Casts\Enum;
+use Tipoff\Support\Contracts\Checkout\Vouchers\VoucherInterface;
+use Tipoff\Support\Contracts\Payment\PaymentInterface;
+use Tipoff\Support\Contracts\Payment\RefundInterface;
 use Tipoff\Support\Models\BaseModel;
 use Tipoff\Support\Traits\HasCreator;
 use Tipoff\Support\Traits\HasPackageFactory;
 use Tipoff\Support\Traits\HasUpdater;
 
-class Refund extends BaseModel
+/**
+ * @property int id
+ * @property Payment payment
+ * @property string refund_number
+ * @property int amount
+ * @property RefundMethod method
+ * @property string|null transaction_number
+ * @property VoucherInterface|null voucher
+ * @property User|null issuer
+ * @property User creator
+ * @property User updater
+ * @property Carbon issued_at
+ * @property Carbon created_at
+ * @property Carbon updated_at
+ * // Raw relations
+ * @property int payment_id
+ * @property int voucher_id
+ * @property int issuer_id
+ * @property int creator_id
+ * @property int updater_id
+ */
+class Refund extends BaseModel implements RefundInterface
 {
     use HasCreator;
     use HasUpdater;
     use HasPackageFactory;
 
-    const METHOD_STRIPE = 'Stripe';
-    const METHOD_VOUCHER = 'Voucher';
-
-    const REFUND_VOUCHER_TYPE_ID = 1;
-
     protected $casts = [
+        'amount' => 'integer',
+        'method' => Enum::class . ':' . RefundMethod::class,
         'issued_at' => 'datetime',
+        'payment_id' => 'integer',
+        'voucher_id' => 'integer',
+        'issuer_id' => 'integer',
+        'creator_id' => 'integer',
+        'updater_id' => 'integer',
     ];
 
     public static function boot()
     {
         parent::boot();
 
-        static::creating(function ($refund) {
-            $refund->generateRefundNumber();
+        static::creating(function (Refund $refund) {
+            $refund->refund_number = static::generateRefundNumber();
         });
 
-        static::saving(function ($refund) {
+        static::saving(function (Refund $refund) {
             Assert::lazy()
                 ->that($refund->payment_id)->notEmpty('A refund must be applied to a payment.')
                 ->that($refund->amount)->notEmpty('A refund must be for an amount.')
                 ->that($refund->amount)->lessOrEqualThan($refund->payment->amount_refundable, 'Please check the payment for the max amount that can be refunded.')
                 ->verifyNow();
         });
 
-        static::created(function ($refund) {
-            $refund
-                ->payment
-                ->generateAmountRefunded()
-                ->save();
+        static::saved(function (Refund $refund) {
+            $payment = $refund->payment;
+            $payment->amount_refunded = static::amountRefunded($payment);
+            $payment->save();
```

Comment:
Since issuing lags creation, this hook needs to be a `saved` hook, not a `created` hook.